### PR TITLE
Add a range of features for augmentation and training, and fix a few bugs

### PR DESCRIPTION
**Pull request by chris-rapson-hectre**
## Motivation

Improve augmentation and training for our model, and hopefully others' models as well.

## Modification

- generalise the heatmap loss calculated in ae_head to have the correct dimensions
- new augmentations: bidirectional flip and rotation by Nx90º
- visualisation to debug augmentations
- add a hook to initially freeze and then unfreeze the backbone, for safer fine tuning
- add a hook for early stopping
- bug fix for singular Hessian matrices (e.g. a constant heatmap)

## BC-breaking (Optional)

No

## Use cases (Optional)

1. associative embedding training failed for me because of a shape mismatch with the keypoint_weights
2. more augmentations are always helpful
3. verifying that augmentations are working as expected helps to debug algorithms and tune hyperparameters
4. pretrained backbones can be vulnerable to large losses during the first few iterations when fine tuning, so it's good to freeze them initially and then allow them to adapt later
5. stop the training when the model starts to overfit to avoid wasting GPU resouce
6. using `pinv` instead of `inv` means the typical case is unchanged and avoids crashing on a pathological edge case.

## Checklist

**Before PR**:

- [y ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ y] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ y] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [y ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [n ] The documentation has been modified accordingly, including docstring or example tutorials.

The only documentation in this PR is as docstrings. Please let me know if you want more.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/datasets/transforms/common_transforms.py [hunk 1]
```diff
@@ -1256,3 +1258,229 @@ def __repr__(self):
...
+        if isinstance(results['img'], list):
+            results['img'] = [np.ascontiguousarray(np.rot90(img, k=k)) for img in results['img']]
+        else:
+            results['img'] = np.ascontiguousarray(np.rot90(results['img'], k=k))
+        
+        if 'img_mask' in results:
+            if isinstance(results['img_mask'], list):
+                results['img_mask'] = [np.ascontiguousarray(np.rot90(mask, k=k)) for mask in results['img_mask']]
+        else:
+            results['img_mask'] = np.ascontiguousarray(np.rot90(results['img_mask'], k=k))
```
Comment (chanul-hectre) [inline mmpose/datasets/transforms/common_transforms.py:1403]: Indent the else block.

File: mmpose/engine/hooks/unfreeze_backbone_hook.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+
+@HOOKS.register_module()
+class UnfreezeBackboneHook(Hook):
+    def __init__(self, unfreeze_epoch, backbone_stages: int = 4):
+        self.unfreeze_epoch = unfreeze_epoch
+        self.backbone_stages = backbone_stages
+
+    def before_run(self, runner: Runner):
+        model = runner.model
```
Comment (chanul-hectre) [inline mmpose/engine/hooks/unfreeze_backbone_hook.py:14]: These lines were added in other repo:

```suggestion
        assert hasattr(runner, 'epoch'), 'Runner must have an epoch attribute'
        assert hasattr(runner.model, 'backbone'), 'Model must have a backbone attribute'
        model = runner.model
```

File: mmpose/engine/hooks/unfreeze_backbone_hook.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            model.backbone._freeze_stages()
+            runner.logger.info('Backbone frozen for initial training phase')
+
+    def before_train_epoch(self, runner: Runner):
+        # Check if we reached the target epoch
+        model = runner.model
+        if runner.epoch >= self.unfreeze_epoch and model.backbone.frozen_stages != -1:
+            runner.logger.info(f'Unfreezing backbone at epoch {runner.epoch}')
+            if hasattr(model, 'module'):  # Handle DistributedDataParallel wrapper
+                model = model.module
```
Comment (chanul-hectre) [inline mmpose/engine/hooks/unfreeze_backbone_hook.py:33]: Move handling of DistributedDataParallel wrapper outside

```suggestion
        model = runner.model
        if hasattr(model, 'module'):  # Handle DistributedDataParallel wrapper
            model = model.module
            
        # Check if we reached the target epoch    
        if runner.epoch >= self.unfreeze_epoch and model.backbone.frozen_stages != -1:
            runner.logger.info(f'Unfreezing backbone at epoch {runner.epoch}')

```

File: tests/test_engine/test_hooks/test_unfreeze_backbone_hook.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+
+class TestUnfreezeBackboneHook(TestCase):
+
+    def test_before_run_freezes_backbone(self):
+        """Verify backbone is frozen at initialization with correct frozen_stages."""
+        hook = UnfreezeBackboneHook(unfreeze_epoch=3)
+
+        model = SimpleNamespace(backbone=DummyBackbone())
+        runner = SimpleNamespace(model=model, logger=Mock())
```
Comment (chanul-hectre) [inline tests/test_engine/test_hooks/test_unfreeze_backbone_hook.py:36]:
```suggestion
        runner = SimpleNamespace(model=model, logger=Mock(), epoch=0)
```

File: tests/test_engine/test_hooks/test_unfreeze_backbone_hook.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    def test_before_train_epoch_unfreezes_on_target_epoch_with_module_wrap(self):
+        """Verify backbone unfreezes at target epoch, handling DistributedDataParallel wrapper."""
+        hook = UnfreezeBackboneHook(unfreeze_epoch=2)
+
+        wrapped_model = SimpleNamespace(backbone=DummyBackbone())
+        runner = SimpleNamespace(
+            model=SimpleNamespace(module=wrapped_model), logger=Mock(), epoch=2
+        )
+
+        hook.before_run(SimpleNamespace(model=wrapped_model, logger=Mock()))
```
Comment (chanul-hectre) [inline tests/test_engine/test_hooks/test_unfreeze_backbone_hook.py:59]:
```suggestion
        hook.before_run(SimpleNamespace(model=wrapped_model, logger=Mock(), epoch=0))
        assert wrapped_model.backbone.frozen_stages == 4
```

File: tests/test_engine/test_hooks/test_visualization_hooks.py [hunk 6]
```diff
@@ -71,3 +77,54 @@ def test_after_test_iter(self):
...
+            assert os.path.isdir(expected_dir)
+
+    def test_before_run_sets_env_even_without_pipeline(self):
+        """Verify env var is set even when pipeline attribute is unavailable."""
+        with tempfile.TemporaryDirectory() as temp_dir:
+            hook = DebugAugmentedSetupHook()
+            runner = SimpleNamespace(
+                work_dir=temp_dir,
+                timestamp="20260330_120000",
+                train_dataloader=SimpleNamespace(dataset=SimpleNamespace()),
```
Comment (chanul-hectre) [inline tests/test_engine/test_hooks/test_visualization_hooks.py:124]:
```suggestion
                train_dataloader=SimpleNamespace(dataset=SimpleNamespace()),
                logger=MagicMock()
```

**Comment by chris-rapson-hectre** [comment]
Sadly, this repository doesn't seem to be responsive any more. I would be happy for somebody to re-open this PR at some point in the future, but I'll close it now so it doesn't show in my open PR list.
